// @slack/events-api/dist/adapter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SlackEventAdapter = exports.errorCodes = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _http = require('http');

var _http2 = _interopRequireDefault(_http);

var _lodash = require('lodash.isstring');

var _lodash2 = _interopRequireDefault(_lodash);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _httpHandler = require('./http-handler');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var debug = (0, _debug2.default)('@slack/events-api:adapter');

var errorCodes = exports.errorCodes = {
  BODY_PARSER_NOT_PERMITTED: 'SLACKADAPTER_BODY_PARSER_NOT_PERMITTED_FAILURE'
};

var SlackEventAdapter = exports.SlackEventAdapter = function (_EventEmitter) {
  _inherits(SlackEventAdapter, _EventEmitter);

  function SlackEventAdapter(signingSecret) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, SlackEventAdapter);

    if (!(0, _lodash2.default)(signingSecret)) {
      throw new TypeError('SlackEventAdapter needs a signing secret');
    }

    var _this = _possibleConstructorReturn(this, (SlackEventAdapter.__proto__ || Object.getPrototypeOf(SlackEventAdapter)).call(this));

    _this.signingSecret = signingSecret;
    _this.includeBody = !!options.includeBody || false;
    _this.includeHeaders = !!options.includeHeaders || false;
    _this.waitForResponse = !!options.waitForResponse || false;

    debug('adapter instantiated - options: %o', {
      includeBody: _this.includeBody,
      includeHeaders: _this.includeHeaders,
      waitForResponse: _this.waitForResponse
    });
    return _this;
  }

  // TODO: options (like https)


  _createClass(SlackEventAdapter, [{
    key: 'createServer',
    value: function createServer() {
      var _this2 = this;

      var path = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '/slack/events';

      // NOTE: this is a workaround for a shortcoming of the System.import() tranform
      return Promise.resolve().then(function () {
        debug('server created - path: %s', path);

        return _http2.default.createServer(_this2.requestListener());
      });
    }
  }, {
    key: 'start',
    value: function start(port) {
      var _this3 = this;

      return this.createServer().then(function (server) {
        return new Promise(function (resolve, reject) {
          _this3.server = server;
          server.on('error', reject);
          server.listen(port, function () {
            return resolve(server);
          });
          debug('server started - port: %s', port);
        });
      });
    }
  }, {
    key: 'stop',
    value: function stop() {
      var _this4 = this;

      return new Promise(function (resolve, reject) {
        if (_this4.server) {
          _this4.server.close(function (error) {
            delete _this4.server;
            if (error) {
              reject(error);
            } else {
              resolve();
            }
          });
        } else {
          reject(new Error('SlackEventAdapter cannot stop when it did not start a server'));
        }
      });
    }
  }, {
    key: 'expressMiddleware',
    value: function expressMiddleware() {
      var middlewareOptions = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var requestListener = this.requestListener(middlewareOptions);
      return function (req, res, next) {
        // If parser is being used, we can't verify request signature
        if (req.body) {
          var error = new Error('Parsing request body prohibits request signature verification');
          error.code = errorCodes.BODY_PARSER_NOT_PERMITTED;
          next(error);
          return;
        }
        requestListener(req, res);
      };
    }
  }, {
    key: 'requestListener',
    value: function requestListener() {
      var middlewareOptions = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      return (0, _httpHandler.createHTTPHandler)(this, middlewareOptions);
    }
  }]);

  return SlackEventAdapter;
}(_events2.default);

/**
 * @alias module:adapter
 */


exports.default = SlackEventAdapter;
